src/category_theory/groupoid.lean

```diff
@@ -48,4 +49,22 @@ def groupoid.iso_equiv_hom : (X ≅ Y) ≃ (X ⟶ Y) :=
 
 end
 
+section
+
+variables {C : Type u} [𝒞 : category.{v} C]
+include 𝒞
+
+/-- A category where every morphism `is_iso` is a groupoid. -/
+def groupoid.of_is_iso (all_is_iso : ∀ {X Y : C} (f : X ⟶ Y), is_iso.{v} f) : groupoid.{v} C :=
+{ inv := λ X Y f, (all_is_iso f).inv }
+
+/-- A category where every morphism has a `trunc` retraction is computably a groupoid. -/
+def groupoid.of_trunc_split_mono (all_split_mono : ∀ {X Y : C} (f : X ⟶ Y), trunc $ split_mono.{v} f) :
+  groupoid.{v} C :=
+groupoid.of_is_iso $ λ X Y f, trunc.rec_on_subsingleton (all_split_mono f) $
+by {introI, exact trunc.rec_on_subsingleton (all_split_mono $ retraction f)
+  (by {introI, exact is_iso.of_mono_retraction})}
```

Comment:
How would you feel if I provided a tactic `trunc_cases` that allowed you to write this proof as
```lean
begin
  apply groupoid.of_is_iso,
  intros X Y f,
  trunc_cases all_split_mono f,
  trunc_cases all_split_mono (retraction f),
  apply is_iso.of_mono_retraction,
end
```
The current proof written in terms of `rec_on_subsingleton` is a bit of a mind-bender to read. :-)

Reply:
That looks good! The current proof going between tactic and term mode is not the most elegant.

Reply:
Let's wait on #2368, then.

Reply:
Okay, that has landed, and I've added a commit here that uses it.